src/lib/Repository/NameSchema/NameSchemaService.php

```diff
@@ -355,6 +270,33 @@ protected function filterNameSchema(string $nameSchema): array
         return [$nameSchema, $groupLookupTable];
     }
 
+    /**
+     * @param array<string, array<string, string>> $tokenValues
+     *
+     * @return array<string, string>
+     */
+    public function buildNames(array $tokenValues, string $nameSchema): array
+    {
+        if (empty($tokenValues)) {
+            throw new UnresolvedTokenNamesException();
+        }
+
+        [$filteredNameSchema, $groupLookupTable] = $this->filterNameSchema($nameSchema);
+        $tokens = $this->extractTokens($filteredNameSchema);
+
+        $names = [];
+        foreach ($tokenValues as $languageCode => $tokenValue) {
+            $name = $filteredNameSchema;
+            foreach ($tokens as $token) {
+                $string = $this->resolveToken($token, $tokenValue, $groupLookupTable);
+                $name = str_replace($token, $string, $name);
+            }
+            $names[$languageCode] = $this->validateNameLength($name);
```

Comment:
However, following the previous comment for the old code, this block could be easily replaced with:
```suggestion
            $names[$languageCode] = $this->validateNameLength(
                str_replace(
                    $tokens,
                    array_map(
                        fn (string $token): string => $this->resolveToken($token, $tokenValue, $groupLookupTable),
                        $tokens
                    ),
                    $filteredNameSchema
                )
            );
```
My primary concern is that nested looping should be less performant than calling `str_replace` directly on arrays, as it was @Steveb-p's point I think. Whether it's more or less readable, it's debatable. Personally I consider nested loops as unreadable.